File: app/src/main/java/fergaral/datetophoto/utils/PhotoUtils.java
package fergaral.datetophoto.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.media.ExifInterface;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.provider.MediaStore;
import android.widget.Toast;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Map;

import fergaral.datetophoto.R;

/**
 * Created by Parejúa on 02/11/2014.
 */
public final class PhotoUtils {

    private Context context;
    private MediaScannerConnection msConn;

    /**
     * Constructor with 1 parameter
     * @param context application context
     */
    public PhotoUtils(Context context) {
        this.context = context;
    }

    /**
     * Method that returns an ArrayList with the photos of the gallery
     * @return ArrayList<String> photos from gallery
     */
    public ArrayList<String> getCameraImages() {
        // Set up an array of the Thumbnail Image ID column we want
        String[] projection = {MediaStore.Images.Media.DATA};

        final Cursor cursorExternal = context.getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                projection,
                null,
                null,
                null);

        final Cursor cursorInternal = context.getContentResolver().query(MediaStore.Images.Media.INTERNAL_CONTENT_URI,
                projection,
                null,
                null,
                null);

        ArrayList<String> result;

        if(cursorExternal != null && cursorInternal == null)
            result = new ArrayList<String>(cursorExternal.getCount());
        else if(cursorExternal != null)
            result = new ArrayList<String>(cursorExternal.getCount() + cursorInternal.getCount());
        else if(cursorInternal != null)
            result = new ArrayList<String>(cursorInternal.getCount());
        else
            result = new ArrayList<String>();

        processPhotosCursor(cursorExternal, result);
        processPhotosCursor(cursorInternal, result);

        return result;
    }

    public List<Thumbnail> getCameraThumbnails() {
        String[] projection = {MediaStore.Images.Thumbnails.DATA, MediaStore.Images.Thumbnails.IMAGE_ID,
                                MediaStore.Images.Thumbnails._ID};

        Cursor cursorExternal = context.getContentResolver().query(MediaStore.Images.Thumbnails.EXTERNAL_CONTENT_URI,
                projection,
                null,
                null,
                null);

        Cursor cursorInternal = context.getContentResolver().query(MediaStore.Images.Thumbnails.INTERNAL_CONTENT_URI,
                projection,
                null,
                null,
                null);

        ArrayList<Thumbnail> cameraThumbnails;

        if(cursorExternal != null && cursorInternal == null)
            cameraThumbnails = new ArrayList<>(cursorExternal.getCount());
        else if(cursorExternal != null)
            cameraThumbnails = new ArrayList<>(cursorExternal.getCount() + cursorInternal.getCount());
        else if(cursorInternal != null)
            cameraThumbnails = new ArrayList<>(cursorInternal.getCount());
        else
            cameraThumbnails = new ArrayList<>();

        processPhotosThumbnailPhotoCursor(cursorExternal, cameraThumbnails);
        processPhotosThumbnailPhotoCursor(cursorInternal, cameraThumbnails);

        return cameraThumbnails;
    }

    private void processPhotosThumbnailPhotoCursor(Cursor cursor, List<Thumbnail> thumbnails) {
        if(cursor == null || thumbnails == null)
            return;

        if(cursor.moveToFirst()) {
            int dataColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Thumbnails.DATA);
            int imageIdColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Thumbnails.IMAGE_ID);
            int idColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Thumbnails._ID);

            do {
                String thumbPath = cursor.getString(dataColumn);
                int imageId = cursor.getInt(imageIdColumn);
                Uri imageURI = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        Integer.toString(cursor.getInt(idColumn)));

                thumbnails.add(new Thumbnail(String.valueOf(imageId), thumbPath, imageURI));
            }while(cursor.moveToNext());
        }

        cursor.close();
    }

    private void processPhotosThumbnailCursor(Cursor cursor, List<String> photosUriList) {
        processImagesCursor(cursor, photosUriList, MediaStore.Images.Thumbnails.DATA);
    }

    private void processPhotosCursor(Cursor cursor, List<String> photosUriList)
    {
        processImagesCursor(cursor, photosUriList, MediaStore.Images.Media.DATA);
    }

    private void processImagesCursor(Cursor cursor, List<String> photosUriList, String imageDataColumn) {
        if(photosUriList == null || cursor == null)
            return;

        if (cursor.moveToFirst()) {
            final int dataColumn = cursor.getColumnIndexOrThrow(imageDataColumn);
            do {
                final String data = cursor.getString(dataColumn);
                //Log.i("data :", data);
                photosUriList.add(data);
            } while (cursor.moveToNext());
        }

        cursor.close();
    }

    /**
     * Method that puts date to the photos of the gallery
     * @param cameraImages ArrayList with the photos of the gallery
     */
    public void processPhotos(ArrayList<String> cameraImages)
    {

        ArrayList<String> galleryImages = cameraImages;

        if(galleryImages != null) {
            int total = 0, actual = 0;
            total = galleryImages.size();


            for (String s : galleryImages) {
                File imgFile = new File(s);
                if (imgFile.exists()) {


                    if (!imgFile.getName().contains("dtp-")) {


                        File imgFileWithDate = new File(imgFile.getParentFile().getAbsolutePath() + "/dtp-" + imgFile.getName());

                        if (!imgFileWithDate.exists()) {


                            BitmapFactory.Options options = new BitmapFactory.Options();
                            //options.inMutable = true;
                            options.inPreferredConfig = Bitmap.Config.ARGB_8888;

                            Bitmap myBitmap = BitmapFactory.decodeFile(imgFile.getAbsolutePath(), options);
                            myBitmap = convertToMutable(myBitmap);

                            String date = "";

                            try {
                                ExifInterface exifInterface = new ExifInterface(imgFile.getAbsolutePath());
                                date = getExifTag(exifInterface, ExifInterface.TAG_DATETIME, imgFile);
                            } catch (IOException e) {
                                e.printStackTrace();
                            }

                            Bitmap bitmap2 = writeDateOnBitmap(myBitmap, date);

                            myBitmap = null;

                            //CapturePhotoUtils.insertImage(getContentResolver(), bitmap2, imgFile.getName() + "-dtp.jpg", "generated using Date To Photo");
                            savePhoto(bitmap2, imgFile.getParentFile().getAbsolutePath(), "dtp-" + imgFile.getName());

                            bitmap2 = null;

                        }
                    }
                }

                actual++;

                imgFile = null;
            }

        }
        else
        {
            /*AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(MyActivity.this);
            alertDialogBuilder.setTitle("No hay fotos");
            alertDialogBuilder.setMessage("No se han encontrado fotos en la memoria de tu dispositivo");
            alertDialogBuilder.setCancelable(true);
            alertDialogBuilder.setPositiveButton("Aceptar", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.cancel();
                }
            });
            alertDialogBuilder.show();

            return null;*/
        }
    }

    /**
     * Method that returns the date associated with a photo, which can be EXIF, or if it's not present, the file date
     * @param exif ExifInterface object
     * @param tag EXIF tag
     * @param imgFile File (used to get file date if EXIF date is not present)
     * @return String date of the photo
     */
    private String getExifTag(ExifInterface exif, String tag, File imgFile) {
        //De la forma: 2014:09:21 13:53:58
        String attribute = exif.getAttribute(tag);

        if(attribute != null) {
            /*String date = attribute.substring(0, attribute.indexOf(" ")); //Ejemplo: 2014:09:21
            StringBuffer buffer = new StringBuffer(date);

            for (int i = 0; i < date.length(); i++) {
                if (date.charAt(i) == ':')
                    buffer.setCharAt(i, '/');
            }

            String dateString = "";

            try {
                Date date1 = new SimpleDateFormat("yyyy-MM-dd").parse(buffer.toString());
                dateString = new SimpleDateFormat("dd-MM-yyyy").format(date1);
            } catch (ParseException e) {
                e.printStackTrace();
            }

            return dateString;*/

            String year = attribute.substring(0, 4);
            String month = attribute.substring(5, 7);
            String day = attribute.substring(8, 10);

            return day + "/" + month + "/" + year;
        }

        Date lastModDate = new Date(imgFile.lastModified());
        Calendar cal = Calendar.getInstance();
        cal.setTime(lastModDate);
        int year = cal.get(Calendar.YEAR);
        int month = cal.get(Calendar.MONTH) + 1;
        int day = cal.get(Calendar.DAY_OF_MONTH);

        return day + "/" + month + "/" + year;
    }

    /**
     * Method that converts an immutable Bitmap to a mutable one
     * @param imgIn immutable Bitmap
     * @return Bitmap the bitmap passed in, but mutable
     */
    private Bitmap convertToMutable(Bitmap imgIn) {
        try {
            //this is the file going to use temporally to save the bytes.
            // This file will not be a image, it will store the raw image data.
            File file = new File(Environment.getExternalStorageDirectory() + File.separator + "temp.tmp");

            //Open an RandomAccessFile
            //Make sure you have added uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"
            //into AndroidManifest.xml file
            RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");

            // get the width and height of the source bitmap.
            int width = imgIn.getWidth();
            int height = imgIn.getHeight();
            Bitmap.Config type = imgIn.getConfig();

            //Copy the byte to the file
            //Assume source bitmap loaded using options.inPreferredConfig = Config.ARGB_8888;
            FileChannel channel = randomAccessFile.getChannel();
            MappedByteBuffer map = channel.map(FileChannel.MapMode.READ_WRITE, 0, imgIn.getRowBytes()*height);
            imgIn.copyPixelsToBuffer(map);
            //recycle the source bitmap, this will be no longer used.
            imgIn.recycle();
            System.gc();// try to force the bytes from the imgIn to be released

            //Create a new bitmap to load the bitmap again. Probably the memory will be available.
            imgIn = Bitmap.createBitmap(width, height, type);
            map.position(0);
            //load it back from temporary
            imgIn.copyPixelsFromBuffer(map);
            //close the temporary file and channel , then delete that also
            channel.close();
            randomAccessFile.close();

            // delete the temp file
            file.delete();

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return imgIn;
    }

    /**
     * Method that puts date on a Bitmap
     * @param b Bitmap without date
     * @param text Date as a string
     * @return Bitmap with date
     */
    public Bitmap writeDateOnBitmap(Bitmap b, String text) {

        float scale = context.getResources().getDisplayMetrics().density;
        Bitmap.Config bitmapConfig = b.getConfig();

        if(bitmapConfig == null) {
            bitmapConfig = Bitmap.Config.ARGB_8888;
        }

        //b = b.copy(bitmapConfig, true);


        Canvas canvas = new Canvas(b);
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(Color.YELLOW);
        paint.setTextSize(b.getHeight() / 20);
        paint.setShadowLayer(1f, 0f, 1f, Color.WHITE);

        Rect bounds = new Rect();
        paint.getTextBounds(text, 0, text.length(), bounds);
        int x = (b.getWidth() - bounds.width()) / 2;
        int y = (b.getHeight() + bounds.height()) / 2;

        x = b.getWidth() - bounds.width();
        y = b.getHeight() - bounds.height();

        canvas.drawText(text, x, y, paint);

        //Toast.makeText(this, String.valueOf(b.getHeight()), Toast.LENGTH_LONG).show();
        //Toast.makeText(this, "Height: "+ String.valueOf(b.getHeight()) + " Width: " + String.valueOf(b.getWidth()), Toast.LENGTH_LONG).show();
        return b;
    }

    /**
     * Method that saves the Bitmap passed in to the device's gallery
     * @param bmp
     * @param basePath
     * @param name
     */
    public void savePhoto(Bitmap bmp, String basePath, String name)
    {
        File imageFileFolder = new File(basePath);
        FileOutputStream out = null;
        File imageFileName = new File(imageFileFolder, name);
        try
        {
            out = new FileOutputStream(imageFileName);
            bmp.compress(Bitmap.CompressFormat.JPEG, 100, out);
            out.flush();
            out.close();
            scanPhoto(imageFileName.toString());
            out = null;
        } catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public void scanPhoto(final String imageFileName)
    {

        msConn = new MediaScannerConnection(context,new MediaScannerConnection.MediaScannerConnectionClient()
        {
            public void onMediaScannerConnected()
            {
                msConn.scanFile(imageFileName, null);
            }
            public void onScanCompleted(String path, Uri uri)
            {
                msConn.disconnect();
            }
        });
        msConn.connect();
    }

    public static ArrayList<String> getFolders(Context ctx)
    {
        ArrayList<String> cameraImages = new PhotoUtils(ctx).getCameraImages();
        ArrayList<String> folderNames = new ArrayList<String>();

        for(String path : cameraImages)
        {
            String parentFolderName = getParentFolderName(path);

            if(!folderNames.contains(parentFolderName))
            {
                folderNames.add(parentFolderName);
            }
        }

        return folderNames;
    }


    public static String getParentFolderName(String path)
    {
        String[] pathSegments = path.split("/");

        return pathSegments[pathSegments.length - 2];
    }


    public static void deletePhoto(Context context, String s)
    {
        File file = new File(s);
        String filePath = file.getAbsolutePath();

        file.delete();
        MediaScannerConnection.scanFile(context, new String[]{filePath}, null, null);
    }

    public static void selectAllFolders(Context context) {
        ArrayList<String> folderNames = PhotoUtils.getFolders(context);
        StringBuilder stringBuilder = new StringBuilder();

        for(String folderName : folderNames) {
            stringBuilder.append(folderName).append(FoldersListPreference.SEPARATOR);
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(context.getString(R.string.pref_folderstoprocess_key), stringBuilder.toString());
        editor.apply();
    }

    public static void selectAllFoldersOnFirstUse(Context context) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);

        if(prefs.getBoolean("firstuse", true))
            selectAllFolders(context);

        SharedPreferences.Editor editor = prefs.edit();
        editor.putBoolean("firstuse", false);
        editor.apply();
    }
}

File: app/src/main/java/fergaral/datetophoto/utils/PhotoUtils.kt
package fergaral.datetophoto.utils

import android.annotation.SuppressLint
import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.graphics.BitmapFactory
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Build
import android.preference.PreferenceManager
import android.provider.MediaStore
import android.util.Log
import androidx.exifinterface.media.ExifInterface
import fergaral.datetophoto.DateToPhoto
import fergaral.datetophoto.R
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

/**
 * Created by Parejúa on 02/11/2014.
 */
class PhotoUtils(private val context: Context) {
    private var msConn: MediaScannerConnection? = null

    val folders: Set<String>
        @SuppressLint("InlinedApi")
        get() {
            val projection = arrayOf(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)

            return queryMediaStore<String>(projection, null, null) { cursorExternal, result ->
                processFoldersCursor(cursorExternal, result)
            }.toSet()
        }

    /**
     * Method that returns an ArrayList with the photos of the gallery
     * @return ArrayList<String> photos from gallery
     */
    @SuppressLint("InlinedApi")
    fun getCameraImages(foldersToProcess: Array<String>): ArrayList<Image> {
        val bucketName = MediaStore.Images.Media.BUCKET_DISPLAY_NAME
        val projection = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            arrayOf(MediaStore.Images.Media.DISPLAY_NAME, bucketName, MediaStore.Images.Media._ID, MediaStore.Images.Media.DATE_ADDED)
        } else {
            arrayOf(MediaStore.Images.Media.DISPLAY_NAME, bucketName, MediaStore.Images.Media._ID, MediaStore.Images.Media.DATE_ADDED, MediaStore.Images.Media.DATA)
        }

        val query = Array(foldersToProcess.size) { "?" }.joinToString(",")
        return queryMediaStore(projection, "$bucketName IN ($query)", foldersToProcess) { cursorExternal, result ->
            processPhotosCursor(cursorExternal, MediaStore.Images.Media.EXTERNAL_CONTENT_URI, result)
        }
    }

    @SuppressLint("InlinedApi")
    fun getCameraImages(): ArrayList<Image> {
        val bucketName = MediaStore.Images.Media.BUCKET_DISPLAY_NAME
        val projection = arrayOf(MediaStore.Images.Media.DISPLAY_NAME, bucketName, MediaStore.Images.Media._ID, MediaStore.Images.Media.DATE_ADDED)

        return queryMediaStore(projection, null, null) { cursorExternal, result ->
            processPhotosCursor(cursorExternal, MediaStore.Images.Media.EXTERNAL_CONTENT_URI, result)
        }
    }

    private fun <T> queryMediaStore(projection: Array<String>, selection: String?, selectionArgs: Array<String>?, callback: (Cursor?, MutableList<T>) -> Unit): ArrayList<T> {
        val cursorExternal = context.contentResolver.query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                projection, selection, selectionArgs,
                MediaStore.MediaColumns.DATE_ADDED + " DESC")

        val result: ArrayList<T> = if (cursorExternal != null)
            ArrayList(cursorExternal.count)
        else
            ArrayList()

        callback(cursorExternal, result)

        return result
    }

    private fun processPhotosCursor(cursor: Cursor?, baseUri: Uri, photosUriList: MutableList<Image>) {
        processImagesCursor(cursor, baseUri, photosUriList)
    }

    @SuppressLint("InlinedApi")
    private fun processImagesCursor(cursor: Cursor?, baseUri: Uri, photosUriList: MutableList<Image>) {
        if (cursor == null)
            return

        if (cursor.moveToFirst()) {
            val idColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID)
            val displayNameColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DISPLAY_NAME)
            val bucketNameColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)
            val dateAddedColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATE_ADDED)
            do {
                val id = cursor.getLong(idColumn)
                val displayName = cursor.getString(displayNameColumn)
                val bucketName = cursor.getString(bucketNameColumn)
                val dateAdded = cursor.getLong(dateAddedColumn)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    photosUriList.add(Image(displayName, bucketName, dateAdded, ContentUris.withAppendedId(baseUri, id)))
                } else {
                    val path = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA))
                    photosUriList.add(Image(displayName, bucketName, dateAdded, ContentUris.withAppendedId(baseUri, id), path))
                }
            } while (cursor.moveToNext())
        }

        cursor.close()
    }

    @SuppressLint("InlinedApi")
    private fun processFoldersCursor(cursor: Cursor?, foldersList: MutableList<String>) {
        if (cursor == null)
            return

        if (cursor.moveToFirst()) {
            val bucketNameColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)
            do {
                val bucketName = cursor.getString(bucketNameColumn)
                if ("Date To Photo" != bucketName) {
                    foldersList.add(bucketName)
                }
            } while (cursor.moveToNext())
        }

        cursor.close()
    }

    /**
     * Method that returns the date associated with a photo, which can be EXIF, or if it's not present, the file date
     * @param exif ExifInterface object
     * @param tag EXIF tag
     * @param imgFile File (used to get file date if EXIF date is not present)
     * @return String date of the photo
     */
    private fun getExifTag(exif: ExifInterface, tag: String, imgFile: File): String {
        //De la forma: 2014:09:21 13:53:58
        val attribute = exif.getAttribute(tag)

        if (attribute != null) {
            /*String date = attribute.substring(0, attribute.indexOf(" ")); //Ejemplo: 2014:09:21
            StringBuffer buffer = new StringBuffer(date);

            for (int i = 0; i < date.length(); i++) {
                if (date.charAt(i) == ':')
                    buffer.setCharAt(i, '/');
            }

            String dateString = "";

            try {
                Date date1 = new SimpleDateFormat("yyyy-MM-dd").parse(buffer.toString());
                dateString = new SimpleDateFormat("dd-MM-yyyy").format(date1);
            } catch (ParseException e) {
                e.printStackTrace();
            }

            return dateString;*/

            val year = attribute.substring(0, 4)
            val month = attribute.substring(5, 7)
            val day = attribute.substring(8, 10)

            return day + "/" + month + "/" + year
        }

        val lastModDate = Date(imgFile.lastModified())
        val cal = Calendar.getInstance()
        cal.time = lastModDate
        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH) + 1
        val day = cal.get(Calendar.DAY_OF_MONTH)

        return day.toString() + "/" + month + "/" + year
    }

    fun scanPhoto(imageFileName: String) {

        msConn = MediaScannerConnection(context, object : MediaScannerConnection.MediaScannerConnectionClient {
            override fun onMediaScannerConnected() {
                msConn!!.scanFile(imageFileName, null)
            }

            override fun onScanCompleted(path: String, uri: Uri) {
                msConn!!.disconnect()
            }
        })
        msConn!!.connect()
    }

    companion object {
        val FIRST_USE_KEY = "firstuse"

        fun getFolders(context: Context): Set<String> = PhotoUtils(context).folders

        fun getParentFolderName(path: String): String {
            val pathSegments = path.split("/".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            return pathSegments[pathSegments.size - 2]
        }

        fun getFileWithDate(fileWithoutDate: String): String {
            val pathSegments = fileWithoutDate.split("/".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
            val name = pathSegments[pathSegments.size - 1]

            return fileWithoutDate.substring(0, fileWithoutDate.length - name.length) + "dtp-" + name
        }

        fun deletePhoto(context: Context, s: String) {
            val file = File(s)
            val filePath = file.absolutePath

            file.delete()
            MediaScannerConnection.scanFile(context, arrayOf(filePath), null, null)
        }

        fun selectAllFolders(context: Context): Set<String> {
            val folderNames = PhotoUtils.getFolders(context)

            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            val editor = prefs.edit()
            editor.putStringSet(context.getString(R.string.pref_folderstoprocess_key), HashSet(folderNames))
            editor.apply()

            return folderNames
        }

        fun selectAllFoldersOnFirstUse(context: Context) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)

            if (prefs.getBoolean(FIRST_USE_KEY, true)) {
                selectAllFolders(context)

                val editor = prefs.edit()
                editor.putBoolean(FIRST_USE_KEY, false)
                editor.apply()
            }
        }

        fun incorrectFormat(image: Image): Boolean {
            return !(image.name.endsWith(".jpg") || image.name.endsWith(".JPG") || image.name.endsWith(".jpeg") ||
                    image.name.endsWith(".JPEG") || image.name.endsWith(".png") || image.name.endsWith(".PNG"))

            //Puede ser que la foto sea, por ejemplo, un JPEG pero que dentro haya texto, no una imagen
            /*BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true; //Solo queremos saber ancho y alto
        BitmapFactory.decodeFile(image, options);

        return options.outWidth == -1 || options.outHeight == -1;*/
        }

        fun getName(image: String): String {
            val pathSegments = image.split("/".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            return pathSegments[pathSegments.size - 1]
        }

        @Throws(IOException::class)
        fun copy(src: File, dst: File) {
            var inStream: FileInputStream? = null
            var outStream: FileOutputStream? = null

            try {
                inStream = FileInputStream(src)
                outStream = FileOutputStream(dst)
                val inChannel = inStream.channel
                val outChannel = outStream.channel
                inChannel.transferTo(0, inChannel.size(), outChannel)
            } finally {
                if (inStream != null)
                    inStream.close()

                if (outStream != null)
                    outStream.close()
            }
        }

        fun isCorrupted(imageFile: File?): Boolean {
            if (imageFile == null)
                return true

            val options = BitmapFactory.Options()
            options.inJustDecodeBounds = true

            BitmapFactory.decodeFile(imageFile.absolutePath, options)

            return options.outWidth == -1 || options.outHeight == -1
        }
    }
}
